### PR TITLE
[JBIDE-22603] - Sometimes multiple OpenShift watch managers are periodically created and finished

### PR DESCRIPTION
**Pull request by jcantrill**

### CODE REVIEW

**Comment by jcantrill** [comment]
@fbricon This partially updates and should supercede #1244 .  It reverts @jeffmaury changes that reverted a change I introduced early where there is 1 listener per kind per project.  We went to this structure because we saw issues where using a single listener per project would receive multiple disconnect notices and we were improperly handling restarting of the watch manager. I would prefer a single listener per project but I don't have the bandwidth to resolve that at the moment.  Additionally, I don't know if this corrects the original issue @jeffmaury was trying to fix, but it does resolve an issue I saw where the same project for 2 separate connections was not being updated properly.

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/IOpenShiftConnection.java [hunk 1]
```diff
@@ -38,6 +38,8 @@
 	
 	<T extends IResource> List<T> getResources(String kind, String namespace);
 
+	<T extends IResource> T getResource(IResource resource);
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/IOpenShiftConnection.java]: That feels pretty confusing from an API perspective. Get a resource from a resource. At the very least this requires some Javadoc

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/WatchManager.java [hunk 2]
```diff
@@ -271,9 +331,9 @@ public void connectionRemoved(IConnection connection) {
 			Connection conn = (Connection)connection;
 			synchronized (watches) {
 				watches.keySet().stream()
-					.filter(p->conn.ownsResource(p))
+					.filter(k->k.connection.equals(conn))
```
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/WatchManager.java]: will that work if the connection token changed?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/IOpenShiftConnection.java (hunk 1 above)
Comment (fbricon) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/IOpenShiftConnection.java]: getResource(...) doesn't seem to be used in this PR

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/WatchManager.java (hunk 2 above)
Comment (jeffmaury) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/WatchManager.java]: connection.equals deals with URL and basic authentication info (so likely will not work with OAuth). Wondering if == should be used

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/IOpenShiftConnection.java (hunk 1 above)
Comment (jeffmaury) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/IOpenShiftConnection.java]: The rationale is that the connection is now used as IOpenShiftConnection and getResource(IResource) was used (in WatchManager) so it was added in the interface definition but I agree I should have copied the Javadoc at least

Comment (jcantrill) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/IOpenShiftConnection.java]: I dont remember this methods purpose?  Is it to refresh a resource?  or to load a resource from basically a predicate?  should we call it something else?

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/WatchManager.java (hunk 2 above)
Comment (jcantrill) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/WatchManager.java]: @fbricon connection should really not consider token during evaluation.  I thought @adietish made changes a while back to address that.  I thought equality was determined to me server + username

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/IOpenShiftConnection.java (hunk 1 above)
Comment (jeffmaury) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/IOpenShiftConnection.java]: It's a refresh

Comment (jcantrill) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/IOpenShiftConnection.java]: I think we should rename it to make it more obvious

Comment (fbricon) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/IOpenShiftConnection.java]: +1 to rename it to refresh(...)

**Comment by jeffmaury** [comment]
testPR

File: plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/WatchManager.java (hunk 2 above)
Comment (adietish) [inline plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/internal/core/WatchManager.java]: @fbricon @jcantrill @jeffmaury this was corrected quite some time ago. Connection#equals does NOT consider authentication info, it just compares username and server-urlhttps://github.com/adietish/jbosstools-openshift/blob/master/plugins/org.jboss.tools.openshift.core/src/org/jboss/tools/openshift/core/connection/Connection.java#L626 You can even see this in the tests: https://github.com/adietish/jbosstools-openshift/blob/master/tests/org.jboss.tools.openshift.test/src/org/jboss/tools/openshift/test/core/connection/ConnectionTest.java#L203

**Comment by fbricon** [comment]
Merged as https://github.com/jbosstools/jbosstools-openshift/commit/d4565ac20d6eeea695890110dc4e3f8ac85a55ba

**Pull request closed without merging**